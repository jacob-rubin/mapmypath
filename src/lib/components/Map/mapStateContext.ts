import type { MapState } from '$lib/state/mapState/mapState.svelte';
import { getContext, setContext } from 'svelte';

const mapStateKey: string = 'mapState';

export function setMapStateContext(mapState: MapState) {
	setContext(mapStateKey, mapState);
}

export function getMapStateContext(): MapState {
	return getContext(mapStateKey) as MapState;
}
