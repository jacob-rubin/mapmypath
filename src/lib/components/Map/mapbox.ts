import mapboxgl from 'mapbox-gl';

import 'mapbox-gl/dist/mapbox-gl.css';

const SOURCE_ID = 'source';
const LAYER_ID = 'layer';

class Mapbox {
	#map: mapboxgl.Map;

	constructor(
		container: string | HTMLElement,
		center: mapboxgl.LngLat = new mapboxgl.LngLat(0, 0),
		zoom: number = 0
	) {
		this.#map = new mapboxgl.Map({
			container,
			accessToken: import.meta.env.VITE_MAPBOX_ACCESS_TOKEN,
			zoom,
			center,
			style: 'mapbox://styles/mapbox/streets-v11'
		});
	}

	remove(): void {
		this.#map.remove();
	}

	isLoaded(): boolean {
		return this.#map.isStyleLoaded();
	}

	awaitLoad(): Promise<void> {
		if (this.isLoaded()) {
			return Promise.resolve();
		} else {
			return new Promise((resolve) => {
				this.#map.on('load', () => {
					resolve();
				});
			});
		}
	}

	initializeStyles(): void {
		this.#map.addSource(SOURCE_ID, {
			type: 'geojson',
			data: {
				type: 'Feature',
				properties: {},
				geometry: {
					type: 'LineString',
					coordinates: []
				}
			}
		});

		this.#map.addLayer({
			id: LAYER_ID,
			source: SOURCE_ID,
			type: 'line',
			layout: {
				'line-join': 'round',
				'line-cap': 'round'
			},
			paint: {
				'line-color': '#888',
				'line-width': 8
			}
		});
	}

	addClickListener(
		callback: (event: mapboxgl.MapMouseEvent) => void
	): void {
		this.#map.on('click', callback);
	}

	addMarker(lngLat: mapboxgl.LngLat): void {
		new mapboxgl.Marker().setLngLat(lngLat).addTo(this.#map);
	}

	getLayer(id: string): mapboxgl.LayerSpecification {
		if (!this.#map.getLayer(id)) {
			throw new Error(`Layer not found: ${id}`);
		}

		return this.#map.getLayer(id)!;
	}

	getSource(id: string): mapboxgl.GeoJSONSourceSpecification {
		if (!this.#map.getSource(id)) {
			throw new Error(`Source not found: ${id}`);
		}

		return this.#map.getSource(id)!.serialize();
	}

	pointToLngLat(point: mapboxgl.Point): mapboxgl.LngLat {
		return this.#map.unproject(point);
	}

	addLineByLngLat(start: mapboxgl.LngLat, end: mapboxgl.LngLat) {
		this.#map.addSource('lineSource', {
			type: 'geojson',
			data: {
				type: 'Feature',
				properties: {},
				geometry: {
					type: 'LineString',
					coordinates: [start.toArray(), end.toArray()]
				}
			}
		});

		this.#map.addLayer({
			id: 'line',
			type: 'line',
			source: 'lineSource',
			layout: {
				'line-join': 'round',
				'line-cap': 'round'
			},
			paint: {
				'line-color': '#888',
				'line-width': 8
			}
		});
	}

	addLineByPoint(start: mapboxgl.Point, end: mapboxgl.Point) {
		this.addLineByLngLat(
			this.pointToLngLat(start),
			this.pointToLngLat(end)
		);
	}

	addMultiLineByLngLat(lngLats: mapboxgl.LngLat[]) {
		this.#map.addSource('multiLineSource', {
			type: 'geojson',
			data: {
				type: 'Feature',
				properties: {},
				geometry: {
					type: 'LineString',
					coordinates: lngLats.map((lngLat) => lngLat.toArray())
				}
			}
		});

		this.#map.addLayer({
			id: 'multiLine',
			type: 'line',
			source: 'multiLineSource',
			layout: {
				'line-join': 'round',
				'line-cap': 'round'
			},
			paint: {
				'line-color': '#888',
				'line-width': 8
			}
		});
	}

	addMultiLineByPoint(points: mapboxgl.Point[]) {
		this.addMultiLineByLngLat(
			points.map((point) => this.pointToLngLat(point))
		);
	}

	renderPath(lngLats: mapboxgl.LngLat[]) {
		const geoJsonSource: mapboxgl.GeoJSONSource =
			this.#map.getSource(SOURCE_ID)!;
		geoJsonSource.setData({
			type: 'Feature',
			properties: {},
			geometry: {
				type: 'LineString',
				coordinates: lngLats.map((lngLat) => lngLat.toArray())
			}
		});
	}
}

export default Mapbox;
