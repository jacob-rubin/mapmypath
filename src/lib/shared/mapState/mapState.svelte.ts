import type { MarkerData } from '$lib/types/markerData';
import type { LngLat } from 'mapbox-gl';

class MapState {
	#markers: MarkerData[] = $state([]); //TODO: consider making this a hash map

	constructor() {
		this.#markers = [];
	}

	addMarker(marker: MarkerData) {
		this.#markers.push(marker);
		console.log($state.snapshot(this.#markers));
	}

	updateMarker(id: number | string, lngLat: LngLat) {
		const marker = this.#markers.find((marker) => marker.id === id);
		if (!marker) {
			throw new Error(`Marker with id ${id} not found`);
		}

		marker.lngLat = lngLat;
	}

	getMarkers() {
		return this.#markers;
	}

	getMarker(id: number): MarkerData {
		const marker: MarkerData | undefined = this.#markers.find(
			(marker) => marker.id === id
		);

		if (!marker) {
			throw new Error(`Marker with id ${id} not found`);
		}

		return marker;
	}

	clear() {
		this.#markers = [];
	}
}

export const mapState = new MapState();
