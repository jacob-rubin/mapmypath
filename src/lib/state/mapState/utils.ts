import type { MapState } from '$lib/state/mapState/mapState.svelte';
import type Mapbox from '$lib/utils/mapbox/mapbox';
import Marker, {
	type MarkerData
} from '$lib/utils/marker/marker.svelte';

export function addMapClickListeners(
	map: Mapbox,
	mapState: MapState
) {
	map.addClickListener((e: mapboxgl.MapMouseEvent) => {
		mapState.addMarker(e.lngLat);
	});
}

// TODO: Clean this up, since don't like having function with 3 inputs
export function addMarkerDragListeners(
	marker: Marker,
	mapState: MapState,
	map: Mapbox
) {
	marker.addDragListener((markerData: MarkerData) => {
		mapState.updateMarker({
			id: markerData.id,
			lngLat: markerData.lngLat
		});
		map.renderPath(mapState.markers.map((m) => m.lngLat));
	});
}
