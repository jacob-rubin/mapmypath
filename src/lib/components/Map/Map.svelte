<script lang="ts">
	import 'mapbox-gl/dist/mapbox-gl.css';
	import { onMount, onDestroy } from 'svelte';
	import { mapState } from '$lib/state/mapState/mapState.svelte';
	import mapboxgl from 'mapbox-gl';
	import Mapbox from '../../utils/mapbox/mapbox';
	import Marker from '$lib/utils/marker/marker.svelte';
	import type { MarkerData } from '$lib/utils/marker/marker.svelte';

	let map: Mapbox;
	let container: HTMLDivElement;

	onMount(async () => {
		let center: mapboxgl.LngLat = new mapboxgl.LngLat(
			-71.224518,
			42.213995
		);
		let zoom: number = 9;

		map = new Mapbox(container, center, zoom);
		await map.awaitLoad();
		map.initializeStyles();

		map.addClickListener((e: mapboxgl.MapMouseEvent) => {
			const marker: Marker = new Marker({
				id: mapState.getMarkers().length, //TODO: This causes bug bc
				lngLat: e.lngLat,
				name: `Stop ${mapState.getMarkers().length + 1}`
			});

			map.addMarker(marker);
			mapState.addMarker(marker);
			map.renderPath(mapState.getMarkers().map((m) => m.lngLat));

			marker.addDragListener((markerData: MarkerData) => {
				mapState.updateMarker({
					id: markerData.id,
					lngLat: markerData.lngLat
				});
				map.renderPath(mapState.getMarkers().map((m) => m.lngLat));
			});
		});
	});

	onDestroy(() => {
		if (map) {
			map.remove();
		}
	});
</script>

<div
	data-testid="map"
	class="h-screen w-screen"
	bind:this={container}
></div>
