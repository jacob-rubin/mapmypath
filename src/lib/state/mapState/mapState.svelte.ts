import { addMapClickListeners } from '$lib/state/mapState/utils';
import type Mapbox from '$lib/utils/mapbox/mapbox';
import Marker from '$lib/utils/marker/marker.svelte';
import type { MarkerData } from '$lib/utils/marker/marker.svelte';

export class MapState {
	#map: Mapbox;
	#markers: Marker[] = $state([]); //TODO: consider making this collection instead of a list

	#counter: number = 0; // TODO: Extrapolate this into a counter class?

	constructor(map: Mapbox) {
		this.#map = map;
		this.#markers = [];

		addMapClickListeners(this.#map, this);
	}

	get markers(): Marker[] {
		return this.#markers;
	}

	set markers(value: Marker[]) {
		this.#markers = value;
	}

	get map(): Mapbox {
		return this.#map;
	}

	getMarkerById(id: number): Marker {
		const marker: Marker | undefined = this.#markers.find(
			(marker) => marker.id === id
		);

		if (!marker) {
			throw new Error(`Marker with id ${id} not found`);
		}

		return marker;
	}

	addMarker(lngLat: mapboxgl.LngLat): Marker {
		// TODO: How can I make this method smaller?
		const marker: Marker = new Marker({
			id: this.#counter++,
			lngLat,
			name: `Stop ${this.#markers.length + 1}`
		});

		marker.addDragListener((markerData: MarkerData) => {
			this.updateMarker({
				id: markerData.id,
				lngLat: markerData.lngLat
			});
			this.#map.renderPath(this.#markers.map((m) => m.lngLat));
		});

		this.#markers.push(marker);
		this.#map.addMarker(marker);
		this.#map.renderPath(this.#markers.map((m) => m.lngLat));

		return marker;
	}

	updateMarker(markerData: MarkerData): void {
		// TODO: Do I need to update the marker lnglat, since it's only updated when dragged?
		//TODO: Can I remove this usage of MarkerData and replace with marker?
		const marker: Marker | undefined = this.#markers.find(
			(marker) => marker.id === markerData.id
		);

		if (!marker) {
			throw new Error(`Marker with id ${markerData.id} not found`);
		}

		marker.lngLat = markerData.lngLat;
	}

	deleteMarker(id: number): Marker {
		const marker: Marker | undefined = this.#markers.find(
			(marker) => marker.id === id
		);

		if (!marker) {
			throw new Error(`Marker with id ${id} not found`);
		}

		this.#markers = this.#markers.filter(
			(marker) => marker.id !== id
		);
		this.#map.deleteMarker(marker);
		this.#map.renderPath(this.#markers.map((m) => m.lngLat));

		return marker;
	}
}
