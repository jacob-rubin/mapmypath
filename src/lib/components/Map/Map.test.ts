import { expect, test } from 'vitest';
import { render } from 'vitest-browser-svelte';
import { fireEvent, screen } from '@testing-library/svelte';
import Map from './Map.svelte';
import { mapState } from '$lib/shared/mapState.svelte';
import '@testing-library/jest-dom/vitest';
import '@testing-library/jest-dom/matchers';

test('it should render the map', async () => {
	const map = render(Map);
	expect(map.getByTestId('map')).toBeDefined();
});

test('it should fill the whole screen', async () => {
	render(Map);

	const map = screen.getByTestId('map');

	const { width, height } = map.getBoundingClientRect();
	expect(width).toBe(window.innerWidth);
	expect(height).toBe(window.innerHeight);
});

test('it should add a marker when the map is clicked', async () => {
	const { container } = render(Map);

	// Simulate a click event on the map
	const mapContainer = container.querySelector('.map-container');
	if (mapContainer) {
		const clickEvent = new MouseEvent('click', {
			clientX: 100,
			clientY: 100,
			bubbles: true,
			cancelable: false
		});
		await fireEvent(mapContainer, clickEvent);

		// Check if a marker element is added to the map
		const marker = container.querySelector('.mapboxgl-marker')!;
		expect(marker).toBeInTheDocument();

		// Check if the marker is at the correct position
		const markerPosition = marker.getBoundingClientRect();
		expect(markerPosition.left).toBeCloseTo(clickEvent.clientX, 1);
		expect(markerPosition.top).toBeCloseTo(clickEvent.clientY, 1);
	}
});

// When map clicked, it should add element to the markers mapState
test('it should add a marker to the mapState when the map is clicked', async () => {
	const { container } = render(Map);

	// Simulate a click event on the map
	const mapContainer = container.querySelector('.map-container');
	if (mapContainer) {
		const clickEvent = new MouseEvent('click', {
			clientX: 100,
			clientY: 100,
			bubbles: true,
			cancelable: false
		});

		expect(mapState.getMarkers()).toHaveLength(0);
		await fireEvent(mapContainer, clickEvent);
		expect(mapState.getMarkers()).toHaveLength(1);
	}
});

// TODO: Get snapshot testing work
