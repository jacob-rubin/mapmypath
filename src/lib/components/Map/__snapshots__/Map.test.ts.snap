// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Map > it matches the snapshot 1`] = `"<div data-testid="map" class="h-screen w-screen mapboxgl-map"><div class="mapboxgl-canary" style="visibility: hidden;"></div><div class="mapboxgl-canvas-container mapboxgl-interactive mapboxgl-touch-drag-pan mapboxgl-touch-zoom-rotate"><canvas class="mapboxgl-canvas" tabindex="0" aria-label="Map" role="region" width="621" height="1344" style="width: 414px; height: 896px;"></canvas></div><div class="mapboxgl-control-container"><div class="mapboxgl-ctrl-top-left"></div><div class="mapboxgl-ctrl-top"></div><div class="mapboxgl-ctrl-top-right"></div><div class="mapboxgl-ctrl-right"></div><div class="mapboxgl-ctrl-bottom-right"><div class="mapboxgl-ctrl mapboxgl-ctrl-attrib mapboxgl-attrib-empty mapboxgl-compact"><button class="mapboxgl-ctrl-attrib-button" type="button" aria-label="Toggle attribution"><span class="mapboxgl-ctrl-icon" aria-hidden="true" title="Toggle attribution"></span></button><div class="mapboxgl-ctrl-attrib-inner"></div></div></div><div class="mapboxgl-ctrl-bottom"></div><div class="mapboxgl-ctrl-bottom-left"><div class="mapboxgl-ctrl" style="display: block;"><a class="mapboxgl-ctrl-logo" target="_blank" rel="noopener nofollow" href="https://www.mapbox.com/" aria-label="Mapbox homepage"></a></div></div><div class="mapboxgl-ctrl-left"></div></div></div>"`;
